zine-renderer.js

```diff
@@ -766,14 +766,93 @@ export class ZineRenderer extends EventTarget {
     const position = layer1.getData('position');
     const quaternion = layer1.getData('quaternion');
     const scale = layer1.getData('scale');
-    
+
     this.transformScene.position.fromArray(position);
     this.transformScene.quaternion.fromArray(quaternion);
     this.transformScene.scale.fromArray(scale);
     this.transformScene.updateMatrixWorld();
 
     this.dispatchEvent(new MessageEvent('transformchange'));
   }
+
+  /**
+   * Get a portion of a zine image by specifying a center,
+   * width and height.
+   *
+   * @param {number} sx Screen-space x coordinate
+   * @param {number} sy Screen-space y coordinate
+   * @param {number} width Width of the image
+   * @param {number} height Height of the image
+   * @return {Promise} Resolves to an image
+   */
+  async getColorImage(
+    sx = 0,
+    sy = 0,
+    width = 128,
+    height = 128,
+  ) {
+    // Get zine image.
+    const {image} = this.sceneMesh.material.uniforms.map.value;
+
+    // Convert screen space coordinates to pixel coordinates.
+    let x = Math.round((sx + 1) / 2 * image.width);
+    let y = Math.round((sy + 1) / 2 * image.height);
+
+    // We will create a bounding box around this coordinate.
+
+    // Adjust the coordinate to ensure that the bounding box
+    // is within the image.
+    if ( x + width / 2 > image.width ) {

```

Comment:
`Math.min(Math.max())` ?

Reply:
I was gonna make a range() util function to clean up all those min/maxes everywhere, will use here

Reply:
done